crates/rethnet_evm/benches/state_ref_basic.rs

```diff
@@ -0,0 +1,94 @@
+use criterion::{black_box, criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion};
+use revm::primitives::AccountInfo;
+use std::str::FromStr;
+
+use rethnet_eth::Address;
+use rethnet_evm::state::{LayeredState, RethnetLayer};
+
+fn make_state(number_of_addresses: u64) -> LayeredState<RethnetLayer> {
+    let mut state = LayeredState::<RethnetLayer>::default();
+    for i in 1..number_of_addresses {
+        use rethnet_evm::state::StateDebug;
+        state
+            .insert_account(
+                Address::from_str(&format!("0x{:0>40x}", i)).unwrap(),
+                AccountInfo::default(),
+            )
+            .unwrap();
+        state.checkpoint().unwrap();
+    }
+    state
+}
+
+const ADDRESS_SCALES: [u64; 4] = [100, 500, 1000, 2000];
+
+fn bench_layered_create(c: &mut Criterion) {
+    for number_of_addresses in ADDRESS_SCALES.iter() {
+        c.bench_with_input(
+            BenchmarkId::new("layered_create_with_{}_addresses", *number_of_addresses),
+            number_of_addresses,
+            |b, number_of_addresses| {
+                b.iter(|| black_box(make_state(*number_of_addresses)));
+            },
+        );
+    }
+}
+
+fn bench_layered_basic(c: &mut Criterion) {
+    for number_of_addresses in ADDRESS_SCALES.iter() {
+        c.bench_with_input(
+            BenchmarkId::new("layered_basic_with_{}_addresses", *number_of_addresses),
+            number_of_addresses,
+            |b, number_of_addresses| {
+                b.iter_batched(
+                    || make_state(*number_of_addresses),
+                    |mut state| {
+                        for i in *number_of_addresses..1 {
+                            use revm::db::State;
+                            black_box(
+                                state
+                                    .basic(Address::from_str(&format!("0x{:0>40x}", i)).unwrap())
+                                    .unwrap(),
+                            );
+                        }
+                    },
+                    BatchSize::LargeInput,
+                )
+            },
+        );
+    }
+}
+
+fn bench_layered_basic_with_clone(c: &mut Criterion) {
+    for number_of_addresses in ADDRESS_SCALES.iter() {
+        let state = make_state(*number_of_addresses);
+        c.bench_with_input(
+            BenchmarkId::new(
+                "layered_basic_with_clone_with_{}_addresses",
+                *number_of_addresses,
+            ),
+            number_of_addresses,
+            |b, number_of_addresses| {
+                b.iter_batched(
+                    || state.clone(),
+                    |mut state| {
+                        for i in *number_of_addresses..1 {
+                            use revm::db::State;
```

Comment:
Could you please move all of the local `use` statements to the top of the file. There is no benefit in Rust to have them locally, in this case.

Use cases where it would be helpful is:
- if you have conflicting naming; or
- if you are including `use` statement in a scope that's behind a feature flag, as you'd also need to put the `use` statement itself behind the feature flag if you do it at the top of the file.

Reply:
I'm sorry I've been stubborn about this, and I realize that there is no benefit *in Rust*, but as a reader of the code I actually find it immensely beneficial. Given that such imported traits have no direct mention in the file, I appreciate having their placement restricted to the scope they're actually needed in. If that scope is the entire file, due to repeated uses within the file, then at the top of the file is the right place. But if the use is only in a single inner scope, then I feel that the placement should be within that single inner scope. (Note that in this particular case I do have a repetition of `revm::db::State` within this file, but that's because I expect one of those usages to go away depending on which version of that function we end up keeping.)

Reply:
A language server provides you that information if you need a reminder though

Reply:
true, but what if for example i'm just browsing the code on github? etc. maybe i'm being overly principled, but it doesn't seem right to *depend* on tooling for readability. in my latest draft all of the `use` are at the top, by virtue of using `SyncState`, so it's a moot point for this PR, but this does remain one of the (very) few smelly things in my limited Rust experience.

Reply:
GitHub these days also offers redirects to type definitions. It seems like you're trying to tailor to a niche while cluttering the code with unnecessary noise for the majority of use cases.